lexibank_backstromnorthernpakistan.py

```diff
@@ -1,30 +1,108 @@
+import attr
+from clldutils.misc import slug
 from clldutils.path import Path
+from clldutils.text import split_text, strip_brackets
 from pylexibank.dataset import Dataset as BaseDataset
+from pylexibank.dataset import Concept, Language
+from tqdm import tqdm
+import csv
+
+
+@attr.s
+class Language(Language):
+    Source_ID = attr.ib(default=None)
+    Name = attr.ib(default=None)
+    Glottolog_Name = attr.ib(default=None)
+    Glottocode = attr.ib(default=None)
+    ISO639P3code = attr.ib(default=None)
+    List_ID = attr.ib(default=None)
+    List_Name = attr.ib(default=None)
+    Family = attr.ib(default=None)
+    Location = attr.ib(default=None)
+    Latitude = attr.ib(default=None)
+    Longitude = attr.ib(default=None)
+    Note = attr.ib(default=None)
+
+@attr.s
+class Concept(Concept):
+    ENGLISH = attr.ib(default=None)
 
 
 class Dataset(BaseDataset):
+    id = 'backstromnorthernpakistan'
     dir = Path(__file__).parent
-    id = "backstromnorthernpakistan"
+    concept_class = Concept
+    language_class = Language
 
     def cmd_download(self, **kw):
-        """
-        Download files to the raw/ directory. You can use helpers methods of `self.raw`, e.g.
-
-        >>> self.raw.download(url, fname)
-        """
         pass
 
     def cmd_install(self, **kw):
-        """
-        Convert the raw data to a CLDF dataset.
+        # read raw lexical data
+        data = []
+        for i in ["A.tsv", "B.tsv", "C.tsv", "D.tsv", "E.tsv", "F.tsv"]:
+            with open(self.dir.joinpath("raw", i).as_posix(), encoding='utf-8') as csvfile:
+                reader = csv.DictReader(csvfile, delimiter="\t")
+                raw_lexemes = [row for row in reader]
+            data.extend(raw_lexemes)
 
-        Use the methods of `pylexibank.cldf.Dataset` after instantiating one as context:
-
-        >>> with self.cldf as ds:
-        ...     ds.add_sources(...)
-        ...     ds.add_language(...)
-        ...     ds.add_concept(...)
-        ...     ds.add_lexemes(...)
-        """
+        # add information to the dataset
         with self.cldf as ds:
-            pass
+            # add sources
```

Comment:
if the names of the header in `etc/languages.tsv` or `etc/languages.csv` are identical with those for your language class, you can in fact just write `ds.add_languages()` here. If you want to say: language['Source_ID'] should be your id, yo write `ds.add_languages(id_factory=lambda x: slug(x['Source_ID']))`

Reply:
There was an issue with identical source IDs; as a result, lexibank recognized 49 varieties instead of 51. I fixed it by adding an additional letter to Rondu Balti (RON_B instead of RON) and Yasin Center (YAS_C instead of YAS). I know that it is not ideal, but I failed to make the code work with 'Name' instead of 'Source_ID'. However, I will find a way to fix it if you think that changing Language_ID to 'Name' would be better than changing the original source IDs.